lib/api/api.go

```diff
@@ -1483,12 +1483,33 @@ func (*service) getDeviceID(w http.ResponseWriter, r *http.Request) {
 
 func (*service) getLang(w http.ResponseWriter, r *http.Request) {
 	lang := r.Header.Get("Accept-Language")
-	var langs []string
+	var langs = make(map[string]float64)
 	for _, l := range strings.Split(lang, ",") {
 		parts := strings.SplitN(l, ";", 2)
-		langs = append(langs, strings.ToLower(strings.TrimSpace(parts[0])))
+		code := strings.ToLower(strings.TrimSpace(parts[0]))
+		langs[code] = 1.0
+		if len(parts) < 2 {
+			continue
+		}
+		weight := strings.ToLower(strings.TrimSpace(parts[1]))
+		if weight[:2] == "q=" {
+			if q, err := strconv.ParseFloat(weight[2:], 32); err != nil {
+				// Completely dismiss entries with invalid weight
+				delete(langs, code)
+			} else {
+				langs[code] = q
+			}
+		}
+	}
+	var orderedLangs = make([]string, 0, len(langs))
+	for code := range langs {
+		orderedLangs = append(orderedLangs, code)
```

Comment:
could add them in the loop above, doesn't really matter much but one loop that defaults the weight to 1.0, tries to parse a weight and skip it it's invalid, then adds to the map and list, would be slightly clearer to me.

Reply:
That's not easily compatible with excluding the ones where parsing the q value fails.  We still need to add it, continue if no quality is assigned, otherwise adjust the q value or remove it again if unparseable.  I'd like to avoid the add-then-remove dance at least on the slice.

Reply:
I meant something like

```
func (*service) getLang(w http.ResponseWriter, r *http.Request) {
	lang := r.Header.Get("Accept-Language")
	weights := make(map[string]float64)
	var langs []string
	for _, l := range strings.Split(lang, ",") {
		parts := strings.SplitN(l, ";", 2)
		code := strings.ToLower(strings.TrimSpace(parts[0]))
		weight := 1.0

		if len(parts) == 2 { // there's a weight part
			if weightStr := strings.ToLower(strings.TrimSpace(parts[1])); strings.HasPrefix(weightStr, "q=") {
				parsedWeight, err := strconv.ParseFloat(weightStr[2:], 32)
				if err != nil {
					// invalid weight; ignore
					continue
				}
				weight = parsedWeight
			}
		}

		langs = append(langs, code)
		weights[code] = weight
	}
	// Reorder by descending q value
	sort.SliceStable(langs, func(i, j int) bool {
		return weights[langs[i]] > weights[langs[j]]
	})
	sendJSON(w, langs)
}
```

though it does get a bit nested in the middle 🤷

Reply:
So... less nesting or two loops?  I prefer the latter, as the operation "order a map by its values" is a standard Go recipe, which resembles the second loop and `sort.SliceStable`.